File: app/src/main/kotlin/org/ninelym/cognitive/embodiment/Unity3DBinding.kt
package org.ninelym.cognitive.embodiment

import org.ninelym.cognitive.CognitiveTensor
import org.ninelym.cognitive.api.AgentType
import org.ninelym.cognitive.api.CognitiveMeshAPI

/**
 * Unity3D Embodiment Binding
 * 
 * Provides integration layer between cognitive mesh and Unity3D game engines.
 * Supports GameObjects, Transforms, Animations, and Unity-specific operations.
 */
class Unity3DBinding(private val meshAPI: CognitiveMeshAPI) {
    
    private val unityAgents = mutableMapOf<String, UnityAgent>()
    
    /**
     * Connect Unity3D client to cognitive mesh
     */
    fun connectUnityClient(
        clientName: String,
        gameObjectCapabilities: List<String>,
        sceneId: String = "MainScene"
    ): Unity3DConnection {
        
        val registrationRequest = org.ninelym.cognitive.api.AgentRegistrationRequest(
            name = "Unity3D-$clientName",
            type = AgentType.UNITY3D,
            capabilities = gameObjectCapabilities + listOf(
                "GameObject", "Transform", "Animation", "Physics", 
                "Rendering", "Audio", "Particles", "UI"
            ),
            autonomyLevel = 0.7f
        )
        
        val response = meshAPI.registerAgent(registrationRequest)
        
        val unityAgent = UnityAgent(
            agentId = response.agentId,
            clientName = clientName,
            sceneId = sceneId,
            gameObjects = mutableMapOf(),
            isConnected = response.success
        )
        
        unityAgents[response.agentId] = unityAgent
        
        return Unity3DConnection(
            agentId = response.agentId,
            success = response.success,
            apiEndpoints = response.apiEndpoints,
            unitySpecificEndpoints = Unity3DEndpoints(
                gameObjects = "/unity/${response.agentId}/gameobjects",
                transforms = "/unity/${response.agentId}/transforms",
                animations = "/unity/${response.agentId}/animations", 
                physics = "/unity/${response.agentId}/physics",
                events = "/unity/${response.agentId}/events"
            )
        )
    }
    
    /**
     * Update GameObject transform data from Unity
     */
    fun updateGameObjectTransform(
        agentId: String,
        gameObjectId: String,
        transform: UnityTransform
    ): Boolean {
        val agent = unityAgents[agentId] ?: return false
        
        // Create cognitive tensor from transform data
        val transformTensor = CognitiveTensor(
            modality = 0.8f, // Spatial/visual modality
            depth = calculateTransformComplexity(transform),
            context = 0.7f,
            salience = calculateTransformSalience(transform),
            autonomyIndex = 0.7f
        )
        
        // Submit as sensor data to mesh
        val sensorRequest = org.ninelym.cognitive.api.SensorDataRequest(
            sensorType = "unity-transform",
            data = transform,
            modality = transformTensor.modality,
            processingDepth = transformTensor.depth,
            contextRelevance = transformTensor.context,
            importance = transformTensor.salience
        )
        
        val response = meshAPI.submitSensorData(agentId, sensorRequest)
        
        if (response.success) {
            // Update local tracking
            agent.gameObjects[gameObjectId] = UnityGameObject(
                id = gameObjectId,
                name = "GameObject_$gameObjectId",
                transform = transform,
                lastUpdate = System.currentTimeMillis()
            )
        }
        
        return response.success
    }
    
    /**
     * Get recommended actions for Unity scene
     */
    fun getSceneRecommendations(agentId: String): List<UnityAction> {
        val actionsResponse = meshAPI.getAvailableActions(agentId)
        
        return actionsResponse.actions.mapNotNull { action ->
            when (action.id) {
                "unity-move" -> UnityAction.MoveGameObject(
                    priority = action.priority,
                    parameters = action.parameters
                )
                "unity-animate" -> UnityAction.TriggerAnimation(
                    priority = action.priority,
                    parameters = action.parameters
                )
                else -> null
            }
        }
    }
    
    /**
     * Handle Unity event (collision, trigger, input, etc.)
     */
    fun handleUnityEvent(
        agentId: String,
        eventType: UnityEventType,
        eventData: Map<String, Any>
    ): UnityEventResponse {
        
        val eventTensor = CognitiveTensor(
            modality = when (eventType) {
                UnityEventType.COLLISION -> 0.9f
                UnityEventType.TRIGGER -> 0.7f
                UnityEventType.INPUT -> 0.8f
                UnityEventType.ANIMATION_COMPLETE -> 0.6f
                UnityEventType.AUDIO -> 0.5f
            },
            depth = 0.8f,
            context = 0.9f, // Events are contextually important
            salience = when (eventType) {
                UnityEventType.COLLISION -> 0.9f
                UnityEventType.INPUT -> 0.8f
                else -> 0.6f
            },
            autonomyIndex = 0.7f
        )
        
        val sensorRequest = org.ninelym.cognitive.api.SensorDataRequest(
            sensorType = "unity-event-${eventType.name.lowercase()}",
            data = eventData,
            modality = eventTensor.modality,
            processingDepth = eventTensor.depth,
            contextRelevance = eventTensor.context,
            importance = eventTensor.salience
        )
        
        val response = meshAPI.submitSensorData(agentId, sensorRequest)
        
        return UnityEventResponse(
            processed = response.success,
            cognitiveRecommendations = response.cognitiveResponse?.recommendations ?: emptyList(),
            suggestedActions = if (response.success) {
                generateEventResponseActions(eventType, eventData)
            } else emptyList()
        )
    }
    
    private fun calculateTransformComplexity(transform: UnityTransform): Float {
        // More complex transforms (rotation, scaling) get higher depth values
        var complexity = 0.3f // Base complexity
        
        if (transform.rotation.x != 0f || transform.rotation.y != 0f || transform.rotation.z != 0f) {
            complexity += 0.3f
        }
        
        if (transform.scale.x != 1f || transform.scale.y != 1f || transform.scale.z != 1f) {
            complexity += 0.2f
        }
        
        return complexity.coerceAtMost(1.0f)
    }
    
    private fun calculateTransformSalience(transform: UnityTransform): Float {
        // Fast-moving or distant objects are more salient
        val magnitude = Math.sqrt(
            (transform.position.x * transform.position.x + 
             transform.position.y * transform.position.y + 
             transform.position.z * transform.position.z).toDouble()
        ).toFloat()
        
        return (magnitude / 100f).coerceIn(0.1f, 1.0f)
    }
    
    private fun generateEventResponseActions(
        eventType: UnityEventType,
        eventData: Map<String, Any>
    ): List<String> {
        return when (eventType) {
            UnityEventType.COLLISION -> listOf(
                "PlayCollisionEffect",
                "UpdateGameObjectState",
                "TriggerAudioFeedback"
            )
            UnityEventType.INPUT -> listOf(
                "ProcessInputAction",
                "UpdatePlayerState", 
                "TriggerInputResponse"
            )
            UnityEventType.TRIGGER -> listOf(
                "ActivateTriggerZone",
                "UpdateTriggerState"
            )
            UnityEventType.ANIMATION_COMPLETE -> listOf(
                "TransitionToNextState",
                "CleanupAnimation"
            )
            UnityEventType.AUDIO -> listOf(
                "ProcessAudioFeedback",
                "UpdateAudioState"
            )
        }
    }
}

/**
 * Unity3D specific data structures
 */

data class Unity3DConnection(
    val agentId: String,
    val success: Boolean,
    val apiEndpoints: org.ninelym.cognitive.api.ApiEndpoints,
    val unitySpecificEndpoints: Unity3DEndpoints
)

data class Unity3DEndpoints(
    val gameObjects: String,
    val transforms: String,
    val animations: String,
    val physics: String,
    val events: String
)

data class UnityAgent(
    val agentId: String,
    val clientName: String,
    val sceneId: String,
    val gameObjects: MutableMap<String, UnityGameObject>,
    var isConnected: Boolean
)

data class UnityGameObject(
    val id: String,
    val name: String,
    val transform: UnityTransform,
    val lastUpdate: Long
)

data class UnityTransform(
    val position: Vector3,
    val rotation: Vector3,
    val scale: Vector3
)

data class Vector3(
    val x: Float,
    val y: Float, 
    val z: Float
)

enum class UnityEventType {
    COLLISION,
    TRIGGER,
    INPUT,
    ANIMATION_COMPLETE,
    AUDIO
}

sealed class UnityAction(val priority: Float, val parameters: Map<String, String>) {
    class MoveGameObject(
        priority: Float,
        parameters: Map<String, String>
    ) : UnityAction(priority, parameters)
    
    class TriggerAnimation(
        priority: Float,
        parameters: Map<String, String>
    ) : UnityAction(priority, parameters)
}

data class UnityEventResponse(
    val processed: Boolean,
    val cognitiveRecommendations: List<String>,
    val suggestedActions: List<String>
)

File: python/helpers/phase4/embodiment/unity3d_binding.py
"""
Unity3D Binding for Cognitive Mesh

Provides seamless integration with Unity3D game engines:
- GameObject transform synchronization
- Physics event handling
- Animation state management
- Real-time scene recommendations
"""

from dataclasses import asdict, dataclass
from enum import Enum
from typing import Any, Dict, List, Optional


# ============================================================================
# Data Classes
# ============================================================================


@dataclass
class Vector3:
    """3D vector for Unity transforms."""
    
    x: float = 0.0
    y: float = 0.0
    z: float = 0.0
    
    @staticmethod
    def zero():
        """Create a zero vector."""
        return Vector3(0, 0, 0)
    
    @staticmethod
    def one():
        """Create a unit vector."""
        return Vector3(1, 1, 1)


@dataclass
class UnityTransform:
    """Unity GameObject transform data."""
    
    position: Vector3
    rotation: Vector3
    scale: Vector3
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "position": asdict(self.position),
            "rotation": asdict(self.rotation),
            "scale": asdict(self.scale),
        }


class UnityEventType(Enum):
    """Unity event types."""
    
    COLLISION = "COLLISION"
    TRIGGER = "TRIGGER"
    INPUT = "INPUT"
    ANIMATION_COMPLETE = "ANIMATION_COMPLETE"
    AUDIO = "AUDIO"


@dataclass
class UnityConnection:
    """Unity client connection information."""
    
    agent_id: str
    success: bool
    message: str
    client_name: str
    capabilities: List[str]
    timestamp: float


# ============================================================================
# Unity3D Binding
# ============================================================================


class Unity3DBinding:
    """
    Unity3D integration binding for cognitive mesh.
    
    Provides GameObject synchronization, event handling,
    and cognitive recommendations for Unity games.
    """
    
    def __init__(self, mesh_api):
        """
        Initialize Unity3D binding.
        
        Args:
            mesh_api: CognitiveMeshAPI instance
        """
        self.mesh_api = mesh_api
        self.transforms: Dict[str, Dict[str, UnityTransform]] = {}
        self.animations: Dict[str, Dict[str, str]] = {}
        
        print("[Unity3DBinding] Initialized")
    
    def connect_unity_client(
        self,
        client_name: str,
        game_object_capabilities: List[str],
        scene_id: str = "MainScene",
    ) -> UnityConnection:
        """
        Connect a Unity3D client to the mesh.
        
        Args:
            client_name: Name of the Unity client
            game_object_capabilities: List of GameObject capabilities
            scene_id: Unity scene identifier
        
        Returns:
            UnityConnection with agent info
        """
        import time
        
        # Register with mesh API
        registration = self.mesh_api.register_agent(
            client_name=client_name,
            capabilities=game_object_capabilities,
            agent_type="unity3d",
        )
        
        if registration.success:
            # Initialize transform and animation tracking
            self.transforms[registration.agent_id] = {}
            self.animations[registration.agent_id] = {}
            
            print(f"[Unity3DBinding] Connected client: {client_name} ({registration.agent_id})")
        
        return UnityConnection(
            agent_id=registration.agent_id,
            success=registration.success,
            message=registration.message,
            client_name=client_name,
            capabilities=game_object_capabilities,
            timestamp=time.time(),
        )
    
    def update_game_object_transform(
        self,
        agent_id: str,
        game_object_id: str,
        transform: UnityTransform,
    ) -> Dict[str, Any]:
        """
        Update GameObject transform data.
        
        Args:
            agent_id: Agent identifier
            game_object_id: GameObject identifier
            transform: Transform data
        
        Returns:
            Response with update status
        """
        if agent_id not in self.transforms:
            return {
                "success": False,
                "error": "Agent not registered",
            }
        
        # Store transform
        self.transforms[agent_id][game_object_id] = transform
        
        # Submit as sensor data
        sensor_data = {
            "gameObjectId": game_object_id,
            "transform": transform.to_dict(),
        }
        
        result = self.mesh_api.submit_sensor_data(
            agent_id=agent_id,
            sensor_type="unity_transform",
            sensor_data=sensor_data,
        )
        
        return {
            "success": result.get("success", False),
            "agent_id": agent_id,
            "game_object_id": game_object_id,
            "transform_updated": True,
        }
    
    def handle_unity_event(
        self,
        agent_id: str,
        event_type: UnityEventType,
        event_data: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Handle Unity event and generate cognitive response.
        
        Args:
            agent_id: Agent identifier
            event_type: Type of Unity event
            event_data: Event data dictionary
        
        Returns:
            Response with recommendations
        """
        if agent_id not in self.transforms:
            return {
                "success": False,
                "error": "Agent not registered",
            }
        
        # Submit event as sensor data
        sensor_data = {
            "eventType": event_type.value,
            "eventData": event_data,
        }
        
        self.mesh_api.submit_sensor_data(
            agent_id=agent_id,
            sensor_type="unity_event",
            sensor_data=sensor_data,
        )
        
        # Generate recommendations based on event type
        recommendations = self._generate_event_recommendations(
            event_type, event_data
        )
        
        return {
            "success": True,
            "agent_id": agent_id,
            "event_type": event_type.value,
            "recommendations": recommendations,
            "cognitive_response": self._get_cognitive_response(event_type),
        }
    
    def get_scene_recommendations(
        self,
        agent_id: str,
        scene_context: Dict[str, Any],
    ) -> Dict[str, Any]:
        """
        Get cognitive recommendations for Unity scene.
        
        Args:
            agent_id: Agent identifier
            scene_context: Current scene context
        
        Returns:
            Recommendations for scene optimization
        """
        if agent_id not in self.transforms:
            return {
                "success": False,
                "error": "Agent not registered",
            }
        
        # Analyze scene context
        object_count = scene_context.get("objectCount", 0)
        active_physics = scene_context.get("activePhysics", 0)
        
        recommendations = []
        
        if object_count > 100:
            recommendations.append({
                "type": "optimization",
                "priority": "high",
                "suggestion": "Consider object pooling for better performance",
            })
        
        if active_physics > 50:
            recommendations.append({
                "type": "physics",
                "priority": "medium",
                "suggestion": "Enable physics LOD for distant objects",
            })
        
        return {
            "success": True,
            "agent_id": agent_id,
            "scene_recommendations": recommendations,
            "performance_score": self._calculate_performance_score(scene_context),
        }
    
    def _generate_event_recommendations(
        self,
        event_type: UnityEventType,
        event_data: Dict[str, Any],
    ) -> List[Dict[str, Any]]:
        """Generate recommendations based on event type."""
        recommendations = []
        
        if event_type == UnityEventType.COLLISION:
            force = event_data.get("force", 0.0)
            if force > 10.0:
                recommendations.append({
                    "action": "trigger_particle_effect",
                    "priority": "high",
                    "params": {"intensity": min(force / 20.0, 1.0)},
                })
        
        elif event_type == UnityEventType.INPUT:
            input_type = event_data.get("inputType", "")
            if input_type == "click":
                recommendations.append({
                    "action": "highlight_object",
                    "priority": "medium",
                    "params": {"duration": 0.5},
                })
        
        elif event_type == UnityEventType.ANIMATION_COMPLETE:
            recommendations.append({
                "action": "transition_to_idle",
                "priority": "low",
                "params": {"blend_time": 0.2},
            })
        
        return recommendations
    
    def _get_cognitive_response(self, event_type: UnityEventType) -> str:
        """Get cognitive response for event type."""
        responses = {
            UnityEventType.COLLISION: "Collision detected - analyzing impact",
            UnityEventType.TRIGGER: "Trigger activated - processing context",
            UnityEventType.INPUT: "Input received - generating response",
            UnityEventType.ANIMATION_COMPLETE: "Animation complete - ready for next state",
            UnityEventType.AUDIO: "Audio event processed - updating scene state",
        }
        return responses.get(event_type, "Event processed")
    
    def _calculate_performance_score(self, scene_context: Dict[str, Any]) -> float:
        """Calculate scene performance score (0-1)."""
        object_count = scene_context.get("objectCount", 0)
        active_physics = scene_context.get("activePhysics", 0)
        draw_calls = scene_context.get("drawCalls", 0)
        
        # Simple scoring based on typical Unity performance metrics
        object_score = max(0.0, 1.0 - (object_count / 500.0))
        physics_score = max(0.0, 1.0 - (active_physics / 200.0))
        draw_call_score = max(0.0, 1.0 - (draw_calls / 1000.0))
        
        return (object_score + physics_score + draw_call_score) / 3.0
